### ISSUE TITLE
Make UnsafeCast less aggressive

### ISSUE THREAD

**Issue by 3flex** [arturbosch/detekt#1601]
## Expected Behavior of the rule
Make UnsafeCast rule less aggressive so it only flags usage of the "unsafe" cast operator when the cast is known to fail:
```kotlin
    val something: String = 3 as String
    val somethingElse: String? = 3 as? String
```
but do not flag usage of the operator when there's not enough information for the compiler to be sure at compile time that the case will fail.

## Context
With type and symbol resolution we can flag usage of the "unsafe" cast operator when we know at compile time that the cast will fail, but ignore usage of the operator when it only *might* fail to avoid false positives.

There are some redundant usages of the "safe" cast operator within detekt that can't currently be changed to the "unsafe" operator without flagging UnsafeCast even when it would be OK, like:
```kotlin
// safe cast to KtValueArgument not necessary since type was already checked
parent is KtValueArgument && (parent as? KtValueArgument)?.isNamed() == true
```
```kotlin
// PomTransaction is a subclass of PomTransactionBase so "unsafe" cast is safe
        override fun runTransaction(transaction: PomTransaction) {
            (transaction as? PomTransactionBase)?.run()
        }
```
```kotlin
// docs say "project.version" returns a string and is never null, so safe cast not needed
override = (project.version as? String)?.endsWith("-SNAPSHOT") == true
```

Changing it to be less aggressive and match behavior of the compiler would help with this.

**Comment by 3flex** [comment]
@arturbosch I'm happy to work on this one but want to check first that you're ok with this approach since it will also reduce the positive detection rate. I think it's better to reduce false positives as much as possible even if the rule starts to miss some real issues. Thoughts?

**Comment by arturbosch** [comment]
@3flex yes I totally agree. This would improve a lot.
